editor/src/components/canvas/controls/select-mode/grid-gap-control-component.tsx

```diff
@@ -0,0 +1,477 @@
+  gapId: string
+  index: number
+  scale: number
+  gapValue: CSSNumberWithRenderedValue
+  axis: Axis
+  onMouseDown: React.MouseEventHandler
+  isDragging: boolean
+  onHandleHoverStartInner: (e: React.MouseEvent, index: number) => void
+  indicatorShown: number | null
+  elementHovered: boolean
+  gapIsHovered: boolean
+  backgroundShown: boolean
+}
+export function GridGapHandle({
+  gapId,
+  index,
+  scale,
+  gapValue,
+  axis,
+  onMouseDown,
+  onHandleHoverStartInner,
+  isDragging,
+  indicatorShown,
+  elementHovered,
+  gapIsHovered,
+  backgroundShown,
+}: GridGapHandleProps) {
+  const { width, height } = handleDimensions(axis, scale)
+  const { hitAreaPadding, paddingIndicatorOffset, borderWidth } = gapControlSizeConstants(
+    DefaultGapControlSizeConstants,
+    scale,
+  )
+  const colorTheme = useColorTheme()
+  const shouldShowIndicator = !isDragging && indicatorShown === index
+  let shouldShowHandle = !isDragging && gapIsHovered
+  // show the handle also if the gap is too narrow to hover
+  if (!gapIsHovered && !backgroundShown) {
+    shouldShowHandle = elementHovered && gapValue.renderedValuePx <= GapHandleGapWidthThreshold
+  }
+  const handleOpacity = gapIsHovered ? 1 : 0.3
+
+  const onHandleHoverStart = React.useCallback(
+    (e: React.MouseEvent) => {
+      onHandleHoverStartInner(e, index)
+    },
+    [onHandleHoverStartInner, index],
+  )
+
+  const rowGapStyles =
+    axis === 'row'
+      ? ({
+          left: '50%',
+          top: '50%',
+          transform: 'translate(-50%, -50%)',
+          position: 'absolute',
+          gridArea: `1/${index + 1}/2/${index + 2}`,
+        } as const)
+      : {}
+  return (
+    <div
+      data-testid={`${GridGapControlHandleTestId}-${gapId}`}
+      style={{
+        visibility: shouldShowHandle ? 'visible' : 'hidden',
+        pointerEvents: 'all',
+        padding: hitAreaPadding,
+        cursor: axis === 'row' ? CSSCursor.GapNS : CSSCursor.GapEW,
+        opacity: handleOpacity,
+        ...rowGapStyles,
+      }}
+      onMouseDown={onMouseDown}
+      onMouseEnter={onHandleHoverStart}
+    >
+      <div
+        style={{
+          position: 'absolute',
+          paddingTop: paddingIndicatorOffset,
+          paddingLeft: paddingIndicatorOffset,
+          pointerEvents: 'none',
+        }}
+      >
+        {when(
+          shouldShowIndicator,
+          <CanvasLabel
+            value={printCSSNumber(
+              gapValue.value ?? cssNumber(gapValue.renderedValuePx, 'px'),
+              null,
+            )}
+            scale={scale}
+            color={colorTheme.brandNeonOrange.value}
+            textColor={colorTheme.white.value}
+          />,
+        )}
+      </div>
+      <PillHandle
+        width={width}
+        height={height}
+        pillColor={colorTheme.brandNeonOrange.value}
+        borderWidth={borderWidth}
+      />
+    </div>
+  )
+}
+
+function handleDimensions(axis: Axis, scale: number): Size {
+  if (axis === 'row') {
```

Comment:
why not a `switch` here?

Reply:
this is just moved code, I wanted to leave most it intact for forensic reasons